sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,314 @@
+
+def gline_strategy(bot, nick):
+    if (known_user := bot.users.get(Identifier(nick))):
+        if hasattr(known_user, "ip"):
+            ip = known_user.ip
+            exemption = get_exemption(ip)
+            if exemption:
+                if "irccloud" in exemption.lower():
+                    # IRCCloud special case: ban uid/sid
+                    return ["gline_username", known_user.user]
+                else: # Fully exempt, so no g-line
+                    return None
+            else: # No exemption
+                return ["gline_ip", ip]
+    else: # Fail safely
+        return None
+
+def gline_or_kill(bot, nick: str, duration: str):
+    if strategy := gline_strategy(bot, nick):
+        if strategy[0] == "gline_ip":
+            gline_ip(bot, strategy[1], duration)
+        elif strategy[0] == "gline_username":
+            gline_username(bot, strategy[1], duration)
+        else:
+            alert(bot, f"Unknown strategy {strategy} for nick {nick}", true)
+            kill_nick(bot, nick) # safest option
+    else:
+        kill_nick(bot, nick) # duration ignored
+
+def protect_chans(bot):
+    if safe_mode:
+        LOGGER.info(f"SAFE MODE: Would protect chans")
+        return
+    for chan in bot.config.emailcheck.protect_chans:
+        bot.write("MODE", chan, "+R")
+
+def malicious_response(bot, nick: str, email):
+    fdrop(bot, nick)
+    add_badmail(bot, email)
+    bot.say(f"You have been temporarily banned from this network because {email.domain} "
+             "has a history of spam or abuse, and/or is a disposable email domain. "
+             "If this is a legitimate domain, contact staff for assistance.",
+             nick.lower())
+    gline_or_kill(bot, nick, bot.config.emailcheck.gline_time)
+    protect_chans(bot)
+    alert(bot, f"ALERT: User {nick} attempted to register a nick with disposable/spam domain {email.domain}!")
+
+def disallow_response(bot, nick: str, email):
+    fdrop(bot, nick)
+    add_badmail(bot, email)
+    bot.say(f"Your registration has been disallowed because {email.domain} appears to be suspicious. "
+             "If this is a legitimate domain, contact staff for assistance.",
+             nick.lower())
+    alert(bot, f"WARNING: User {nick} attempted to register a nick with suspicious domain {email.domain}.")
+
+def fetch_IPQS_email_score(
+    email_addr: str,
+    key: str,
+    fast: bool = True
+    ) -> tuple[float, bool, bool]: #score, disposable, has recent abuse flag set
+    '''Perform lookup on a specific email adress using ipqualityscore.com'''
+    email_str = urllib.parse.quote(email_addr)
+    faststr = str(bool(fast)).lower() #lower + handle None and other garbage
+    params = urllib.parse.urlencode({'fast': faststr})
+    with urllib.request.urlopen(
+        f"https://ipqualityscore.com/api/json/email/{key}/{email_str}?{params}") as url:
+            data = json.loads(url.read().decode())
+            LOGGER.debug(data)
```

Comment:
The response might not be clear that it's coming from a third-party API, I would prefix it with `IP query:` or something

Reply:
The method is getting tossed, along with IPQS, but I'll keep that in mind for the replacement's logging